node/lifecycle_test.go

```diff
@@ -0,0 +1,211 @@
+}
+
+func TestPodLifecycle(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	newLogger := logruslogger.FromLogrus(logrus.NewEntry(logrus.StandardLogger()))
+	logrus.SetLevel(logrus.DebugLevel)
+
+	// Right now, new loggers that are created from spans are broken since log.L isn't set.
+	ctx = log.WithLogger(ctx, newLogger)
+
+	// Create the fake client.
+	client := fake.NewSimpleClientset()
+
+	// This is largely copy and pasted code from the root command
+	podInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(
+		client,
+		informerResyncPeriod,
+		kubeinformers.WithNamespace(testNamespace),
+	)
+	podInformer := podInformerFactory.Core().V1().Pods()
+
+	scmInformerFactory := kubeinformers.NewSharedInformerFactory(client, informerResyncPeriod)
+
+	eb := record.NewBroadcaster()
+	eb.StartLogging(log.G(ctx).Infof)
+	eb.StartRecordingToSink(&corev1client.EventSinkImpl{Interface: client.CoreV1().Events(testNamespace)})
+	fakeRecorder := &fakeDiscardingRecorder{}
+
+	secretInformer := scmInformerFactory.Core().V1().Secrets()
+	configMapInformer := scmInformerFactory.Core().V1().ConfigMaps()
+	serviceInformer := scmInformerFactory.Core().V1().Services()
+
+	mockProvider, err := mock.NewMockProviderMockConfig(mock.MockConfig{}, testNodeName, "linux", "1.2.3.4", 0)
+	assert.NilError(t, err)
+	config := PodControllerConfig{
+		PodClient:         client.CoreV1(),
+		PodInformer:       podInformer,
+		EventRecorder:     fakeRecorder,
+		Provider:          mockProvider,
+		ConfigMapInformer: configMapInformer,
+		SecretInformer:    secretInformer,
+		ServiceInformer:   serviceInformer,
+	}
+
+	pc, err := NewPodController(config)
+	assert.NilError(t, err)
+
+	p := corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "my-pod",
+			Namespace:       testNamespace,
+			UID:             "4f20ff31-7775-11e9-893d-000c29a24b34",
+			ResourceVersion: "100",
+		},
+		Spec: corev1.PodSpec{
+			NodeName: testNodeName,
+		},
+		Status: corev1.PodStatus{
+			Phase: corev1.PodPending,
+		},
+	}
+
+	listOptions := metav1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("metadata.name", p.ObjectMeta.Name).String(),
+	}
+
+	watchErrCh := make(chan error)
+
+	watcher, err := client.CoreV1().Pods(testNamespace).Watch(listOptions)
+	assert.NilError(t, err)
+	go func() {
+		_, watchErr := watchutils.UntilWithoutRetry(ctx, watcher,
+			// Wait for the pod to be created
+			// TODO(Sargun): Make this "smarter" about the status the pod is in.
+			func(ev watch.Event) (bool, error) {
+				pod := ev.Object.(*corev1.Pod)
+				return pod.Name == p.ObjectMeta.Name, nil
+			})
+
+		watchErrCh <- watchErr
+	}()
+
+	go podInformerFactory.Start(ctx.Done())
+	go scmInformerFactory.Start(ctx.Done())
+
+	_, e := client.CoreV1().Pods(testNamespace).Create(&p)
```

Comment:
Please add comments for the major test sections / assertions.  e.g. `// CREATE POD`